File: publish/figuremark/figuremark.rb
# FigureMark, by Matt Gemmell ~ https://mattgemmell.scot/figuremark
# Jekyll plugin. Install: _plugins/figuremark.rb

# Set figuremark: true in a post/page's YAML front-matter to enable processing.
# You can also set figuremark: true in your site's _config.yml file to enable it globally.

class FMAttributes
  # Directives
  FIG_NUM_FORMAT = 'fig-num-format'
  EMPTY_CAPTIONS = 'empty-captions'
  CAPTION_BEFORE = 'caption-before'
  LINK_CAPTION = 'link-caption'
  RETAIN_BLOCK = 'retain-block'

  # Magic
  SHARED_CLASS = "figuremark"
  DIRECTIVE_PREFIX = ":"
  REMOVE_TOKEN = "-#{DIRECTIVE_PREFIX}"

  attr_accessor :tag_id, :classes, :tag_attrs, :directives

  def initialize(raw_string=nil)
    @tag_id = nil
    @classes = [SHARED_CLASS]
    @tag_attrs = {}
    @directives = {}

    if raw_string
      # Parse attribute string like '.class #id key=val'.
      pattern = /([.#][\w:-]+|[\w:-]+=(?:"[^"]*"|'[^']*'|[^\s]*)|[\w\-\.]+)/
      raw_string.scan(pattern).flatten.each do |item|
        if item.start_with?('.') && !@classes.include?(item[1..-1])
          @classes << item[1..-1]
        elsif item.start_with?('#')
          @tag_id = item[1..-1]
        elsif item.include? "="
          key, val = item.split('=', 2)
          val ||= ""
          dest = @tag_attrs
          this_key = key
          if key.start_with?(DIRECTIVE_PREFIX)
            dest = @directives
            this_key = this_key[DIRECTIVE_PREFIX.length..-1]
          end
          dest[this_key] = val.gsub(/\A['"]|['"]\z/, '')
        else
        	split_classes = item.split(".")
        	split_classes.each do |this_class|
        		if !@classes.include?(this_class)
        			@classes << this_class
        		end
        	end
        end
      end
    end
  end

  def to_s
    attr_str = ""
    attr_str += " id=\"#{@tag_id}\"" if @tag_id
    attr_str += " class=\"#{@classes.join(' ')}\"" if @classes.size > 0
    @tag_attrs.each { |k, v| attr_str += " #{k}=\"#{v}\"" }
    attr_str
  end

  def update(new_attrs)
    # OVERWRITE our own values if necessary, process removals
    if new_attrs.tag_id && new_attrs.tag_id.start_with?(REMOVE_TOKEN)
      if @tag_id == new_attrs.tag_id[REMOVE_TOKEN.length..-1] || new_attrs.tag_id == REMOVE_TOKEN
        @tag_id = nil
      end
    else
      @tag_id = new_attrs.tag_id if new_attrs.tag_id
    end

    new_attrs.classes.each do |new_class|
      if new_class == REMOVE_TOKEN
        @classes.clear
      elsif new_class.start_with?(REMOVE_TOKEN)
        @classes.delete(new_class[REMOVE_TOKEN.length..-1])
      elsif !@classes.include?(new_class)
        @classes << new_class
      end
    end

    new_attrs.tag_attrs.each do |k, v|
      if v == REMOVE_TOKEN
        @tag_attrs.delete(k)
      elsif k == REMOVE_TOKEN
        @tag_attrs.clear
      else
        @tag_attrs[k] = v
      end
    end

    new_attrs.directives.each do |k, v|
      if v == REMOVE_TOKEN
        @directives.delete(k)
      elsif k == REMOVE_TOKEN
        @directives.clear
      else
        @directives[k] = v
      end
    end
  end

  def incorporate(new_attrs)
    # Let OWN values take precedence, no removals
    @tag_id ||= new_attrs.tag_id
    new_attrs.classes.each { |cls| @classes << cls unless @classes.include?(cls) }
    @tag_attrs = new_attrs.tag_attrs.merge(@tag_attrs)
    @directives = new_attrs.directives.merge(@directives)
  end
end


Jekyll::Hooks.register [:documents, :pages], :pre_render do |doc|
  # Only process Markdown files.
  ext = doc.respond_to?(:extname) ? doc.extname : File.extname(doc.relative_path)
  next unless ext == '.md' || ext == '.markdown'
	
	proceed = true
	if doc.data.key?('figuremark')
		proceed = (doc.data['figuremark'] == true)
	else
		proceed = (doc.site.config['figuremark'] == true)
	end
	next unless proceed
	
	def convert(doc)
		text = doc.content
		
		fm_globals_pattern = /(?mi)^\{figure(?:mark)?\s*([^\}]*)\}\s*?$/
		figure_block_pattern = /(?mi)(?<!<!--\n)^(`{3,}|~{3,})\s*figure(?:mark)?(\s+[^\{]+?)?\s*(?:\{([^\}]*?)\})?\s*$\n([\s\S\n]*?)\n\1\s*?$/
		figure_span_pattern = /(?<!\\)\[(.+?)(?<!\\)\]\{([^\}]+?)\}|\{([\d.-]+)\}/
		marks_map = {
			"+" => "insert",
			"-" => "remove",
			"/" => "comment",
			">" => "result",
			"!" => "highlight"
		}
	
		figure_number = 0
		figs_processed = 0
		block_pattern_obj = Regexp.new(figure_block_pattern)
		span_pattern_obj = Regexp.new(figure_span_pattern)
		last_fig_end = 0
		global_attrs = FMAttributes.new
	
		# Find any FigureMark blocks needing rewritten.
		text_out = text.dup
		while (block_match = block_pattern_obj.match(text_out, last_fig_end))
			block_title = block_match[2] ? block_match[2].strip : ""
			block_attributes = block_match[3]
			processed_block = block_match[4]
	
			# Process any embedded figure-marking spans.
			last_span_end = 0
			while (span_match = span_pattern_obj.match(processed_block, last_span_end))
				processed_span = ""
				bracketed_text = span_match[1] || ""
				if span_match[3]
					# Reference number without bracketed span.
					ref_num = span_match[3]
					processed_span = "<span class=\"#{FMAttributes::SHARED_CLASS} reference reference-#{ref_num}\">#{ref_num}</span>"
				elsif marks_map.has_key?(span_match[2])
					# Known directive span.
					css_class = marks_map[span_match[2]]
					processed_span = "<span class=\"#{FMAttributes::SHARED_CLASS} #{css_class}\">#{bracketed_text}</span>"
				else
					# Parse as an attribute list.
					span_attrs = FMAttributes.new(span_match[2])
					processed_span = "<span#{span_attrs}>#{bracketed_text}</span>"
				end
				replace_start = span_match.begin(0)
				replace_end = span_match.end(0)
				processed_block = processed_block[0...replace_start] + processed_span + processed_block[replace_end..-1]
				last_span_end = replace_start + processed_span.length
			end
	
			# Sync figure number with any intervening non-FigureMark figures.
			between_text = text_out[last_fig_end...block_match.begin(0)]
			other_figures = between_text.to_enum(:scan, /(?m)<figure[^>]*>.+?<\/figure>/).map { Regexp.last_match }
			figure_number += other_figures.length if other_figures
			figure_number += 1
	
			# Remove escaping backslashes from brackets and braces.
			processed_block.gsub!(/(?<!\\)\\([\[\]\{\}\\])/, '\1')
			processed_block = "<div class=\"figure-content\">#{processed_block}</div>"
			attrs = FMAttributes.new(block_attributes)
			attrs.tag_id ||= "figure-#{figure_number}"
			attrs.tag_attrs['data-fignum'] = figure_number.to_s
	
			# Handle intervening globals blocks.
			pre_match_text = text_out[0...block_match.begin(0)]
			pre_match_orig_len = pre_match_text.length
			globals_matches = []
			pre_match_text.scan(fm_globals_pattern) do
				m = Regexp.last_match
				globals_matches << m
				global_attrs.update(FMAttributes.new(m[1]))
			end
	
			# Remove globals blocks in reverse, to preserve offsets as we go.
			globals_matches.reverse.each do |m|
				pre_match_text = pre_match_text[0...m.begin(0)] + pre_match_text[m.end(0)..-1]
			end
	
			# Account for changed offsets in block_match due to removal of globals blocks.
			pre_match_delta = pre_match_orig_len - pre_match_text.length
	
			# Incorporate global attributes, overriding with local values.
			attrs.incorporate(global_attrs)
	
			# Enact directives.
			fig_num_format = (attrs.directives[FMAttributes::FIG_NUM_FORMAT] || "Fig. #").gsub('#', figure_number.to_s)
			empty_captions = (attrs.directives[FMAttributes::EMPTY_CAPTIONS] || "true") == "true"
			caption_before = (attrs.directives[FMAttributes::CAPTION_BEFORE] || "true") == "true"
			link_caption = attrs.directives[FMAttributes::LINK_CAPTION] || "num"
			retain_block = attrs.directives[FMAttributes::RETAIN_BLOCK] || "none"
	
			if block_title != "" || empty_captions
				link_tag = "<a href=\"##{attrs.tag_id}\">"
				caption_string = "<figcaption><span class=\"figure-number\">#{link_tag}#{fig_num_format}</a></span><span class=\"figure-title\">#{block_title}</span></figcaption>"
				if link_caption == "title"
					caption_string = "<figcaption><span class=\"figure-number\">#{fig_num_format}</span><span class=\"figure-title\">#{link_tag}#{block_title}</a></span></figcaption>"
				elsif link_caption == "all"
					caption_string = "<figcaption>#{link_tag}<span class=\"figure-number\">#{fig_num_format}</span><span class=\"figure-title\">#{block_title}</span></a></figcaption>"
				elsif link_caption == "none"
					caption_string = "<figcaption><span class=\"figure-number\">#{fig_num_format}</span><span class=\"figure-title\">#{block_title}</span></figcaption>"
				end
	
				if caption_before
					processed_block = "#{caption_string}\n#{processed_block}"
				else
					processed_block = "#{processed_block}\n#{caption_string}"
				end
			end
	
			processed_block = "<figure#{attrs}>#{processed_block}</figure>"
	
			if retain_block == "comment"
				processed_block = "<!--\n#{block_match[0]}\n-->\n\n#{processed_block}"
			elsif retain_block == "indent"
				lines = block_match[0].lstrip.lines
				indented = lines.map { |l| "\t#{l}" }.join
				processed_block = "#{indented}\n\n#{processed_block}"
			end
	
			last_fig_end = block_match.begin(0) + processed_block.length - pre_match_delta
			text_out = pre_match_text + processed_block + text_out[block_match.end(0)..-1]
			figs_processed += 1
		end
	
		if figs_processed > 0
			puts "Processed #{figs_processed} FigureMark blocks in \"#{doc.data['title']}\" (#{doc.path})."
		else
			#puts "No FigureMark blocks found in \"#{doc.data['title']}\" (#{doc.path})."
		end
		
		text_out
	end
	
	doc.content = convert(doc)
end

File: publish/figuremark/figuremark.py

# FigureMark, by Matt Gemmell ~ https://mattgemmell.scot/figuremark


import re


class FMAttributes:
	# Directives
	fig_num_format = f"fig-num-format"
	empty_captions = f"empty-captions"
	caption_before = f"caption-before"
	link_caption = f"link-caption"
	retain_block = f"retain-block"
	
	# Magic
	shared_class = "figuremark"
	directive_prefix = ":"
	remove_token = f"-{directive_prefix}"
	
	def __init__(self, raw_string=None):
		self.tag_id = None
		self.classes = [FMAttributes.shared_class]
		self.tag_attrs = {}
		self.directives = {}
		
		if raw_string:
			# Parse attribute string like '.class #id key=val'.
			pattern = re.compile(r'([.#][\w:-]+|[\w:-]+=(?:"[^"]*"|\'[^\']*\'|[^\s]*)|[\w\-\.]+)')
			for match in pattern.findall(raw_string):
				item = match
				if item.startswith('.') and not item[1:] in self.classes:
					self.classes.append(item[1:])
				elif item.startswith('#'):
					self.tag_id = item[1:]
				elif "=" in item:
					key, _, val = item.partition('=')
					val = val or ""
					dest = self.tag_attrs
					this_key = key
					if key.startswith(FMAttributes.directive_prefix):
						dest = self.directives
						# Trim off the directive_prefix.
						this_key = this_key[len(FMAttributes.directive_prefix):]
					dest[this_key] = val.strip('"\'')
				else:
					split_classes = item.split(".")
					for this_class in split_classes:
						if not this_class in self.classes:
							self.classes.append(this_class)
	
	def __str__(self):
		attr_str = ""
		if self.tag_id:
			attr_str += f' id="{self.tag_id}"'
		if len(self.classes) > 0:
			attr_str += f' class="{" ".join(self.classes)}"'
		for k, v in self.tag_attrs.items():
			attr_str += f' {k}="{v}"'
		return attr_str
	
	def update(self, new_attrs):
		# Update self with values from new_attrs, OVERWRITING our own values if necessary.
		# This method also appropriately processes removals with the remove_token.
		
		if new_attrs.tag_id and new_attrs.tag_id.startswith(FMAttributes.remove_token):
			if self.tag_id == new_attrs.tag_id[len(FMAttributes.remove_token):] or new_attrs.tag_id == FMAttributes.remove_token:
				self.tag_id = None
		else:
			self.tag_id = new_attrs.tag_id
		
		for new_class in new_attrs.classes:
			if new_class == FMAttributes.remove_token:
				self.classes.clear()
			elif new_class.startswith(FMAttributes.remove_token):
				if new_class[len(FMAttributes.remove_token):] in self.classes:
					self.classes.remove(new_class[len(FMAttributes.remove_token):])
			elif not new_class in self.classes:
				self.classes.append(new_class)
		
		for k, v in new_attrs.tag_attrs.items():
			if v == FMAttributes.remove_token:
				if k in self.tag_attrs:
					del self.tag_attrs[k]
			elif k == FMAttributes.remove_token:
				self.tag_attrs.clear()
			else:
				self.tag_attrs[k] = v
		
		for k, v in new_attrs.directives.items():
			if v == FMAttributes.remove_token:
				if k in self.directives:
					del self.directives[k]
			elif k == FMAttributes.remove_token:
				self.directives.clear()
			else:
				self.directives[k] = v
	
	def incorporate(self, new_attrs):
		# Update self with values from new_attrs, letting our OWN values take precedence.
		# This method does NOT process removals.
		
		if not self.tag_id:
			self.tag_id = new_attrs.tag_id
		
		for new_class in new_attrs.classes:
			if not new_class in self.classes:
				self.classes.append(new_class)
		
		temp = dict(new_attrs.tag_attrs)
		temp.update(self.tag_attrs)
		self.tag_attrs = temp
		
		temp = dict(new_attrs.directives)
		temp.update(self.directives)
		self.directives = temp


def convert(text):
	fm_globals_pattern = r"(?mi)^(?:\{figure(?:mark)?\s*([^\}]*)\})\s*?$"
	figure_block_pattern = r"(?mi)(?<!<!--\n)^(`{3,}|~{3,})\s*figure(?:mark)?(\s+[^\{]+?)?\s*(?:\{([^\}]*?)\})?\s*$\n([\s\S\n]*?)\n\1\s*?$"
	figure_span_pattern = r"(?<!\\)\[(.+?)(?<!\\)\]\{([^\}]+?)\}|\{([\d.-]+)\}"
	marks_map = {	"+": "insert",
								"-": "remove",
								"/": "comment",
								">": "result",
								"!": "highlight"}
	figure_number = 0
	figs_processed = 0
	block_pattern_obj = re.compile(figure_block_pattern)
	span_pattern_obj = re.compile(figure_span_pattern)
	last_fig_end = 0
	global_attrs = FMAttributes()
	
	# Find any FigureMark blocks needing rewritten.
	block_match = block_pattern_obj.search(text, last_fig_end)
	while block_match:
		block_title = block_match.group(2)
		if block_title:
			block_title = block_title.strip()
		block_attributes = block_match.group(3) 
		processed_block = block_match.group(4)
		
		# Process any embedded figure-marking spans.
		last_span_end = 0
		span_match = span_pattern_obj.search(processed_block, last_span_end)
		while span_match:
			processed_span = ""
			bracketed_text = span_match.group(1) if span_match.group(1) else ""
			if span_match.group(3):
				# Reference number without bracketed span.
				ref_num = span_match.group(3)
				processed_span = f'<span class="{FMAttributes.shared_class} reference reference-{ref_num}">{ref_num}</span>'
				
			elif span_match.group(2) in marks_map:
				# Known directive span.
				css_class = marks_map[span_match.group(2)]
				processed_span = f'<span class="{FMAttributes.shared_class} {css_class}">{bracketed_text}</span>'
				
			else:
				# Parse as an attribute list.
				span_attrs = FMAttributes(span_match.group(2))
				processed_span = f'<span{span_attrs}>{bracketed_text}</span>'
			
			last_span_end = span_match.start() + len(processed_span)
			processed_block = processed_block[:span_match.start()] + processed_span + processed_block[span_match.end():]
			span_match = span_pattern_obj.search(processed_block, last_span_end)
		
		# Sync figure number with any intervening non-FigureMark figures.
		other_figures = re.findall(r"(?sm)<figure[^>]*>.+?</figure>", text[last_fig_end:block_match.start()])
		if other_figures:
			figure_number += len(other_figures)
		figure_number += 1
		
		# Assemble a suitable pre-formatted figure block.
		
		# Remove escaping backslashes from brackets and braces.
		processed_block = re.sub(r"(?<!\\)\\([\[\]\{\}\\])", r"\1", processed_block)
		processed_block = f"<div class=\"figure-content\">{processed_block}</div>"
		attrs = FMAttributes(block_attributes)
		if not attrs.tag_id:
			attrs.tag_id = f"figure-{figure_number}"
		attrs.tag_attrs['data-fignum'] = f'{figure_number}'
		
		# Handle intervening globals blocks.
		pre_match_text = text[:block_match.start()]
		pre_match_orig_len = len(pre_match_text)
		globals_matches = []
		
		# Update global_attrs with each set of found globals, in order.
		for globals_match in re.finditer(fm_globals_pattern, pre_match_text):
			globals_matches.append(globals_match)
			global_attrs.update(FMAttributes(globals_match.group(1)))
				
		# Remove globals blocks in reverse, to preserve offsets as we go.
		for globals_match in reversed(globals_matches):
			pre_match_text = pre_match_text[:globals_match.start()] + pre_match_text[globals_match.end():]
		
		# Account for changed offsets in block_match due to removal of globals blocks.
		pre_match_delta = pre_match_orig_len - len(pre_match_text)
		
		# Incorporate global attributes, overriding with local values.
		attrs.incorporate(global_attrs)
		
		# Enact directives.
		fig_num_format = attrs.directives.get(FMAttributes.fig_num_format, "Fig. #").replace("#", str(figure_number))
		empty_captions = (attrs.directives.get(FMAttributes.empty_captions, "true") == "true")
		caption_before = (attrs.directives.get(FMAttributes.caption_before, "true") == "true")
		link_caption = attrs.directives.get(FMAttributes.link_caption, "num") # | title | all | none
		retain_block = attrs.directives.get(FMAttributes.retain_block, "none") # | comment | indent
		
		if block_title != "" or empty_captions:
			link_tag = f"<a href=\"#{attrs.tag_id}\">"
			caption_string = f"<figcaption><span class=\"figure-number\">{link_tag}{fig_num_format}</a></span><span class=\"figure-title\">{block_title}</span></figcaption>"
			if link_caption == "title":
				caption_string = f"<figcaption><span class=\"figure-number\">{fig_num_format}</span><span class=\"figure-title\">{link_tag}{block_title}</a></span></figcaption>"
			elif link_caption == "all":
				caption_string = f"<figcaption>{link_tag}<span class=\"figure-number\">{fig_num_format}</span><span class=\"figure-title\">{block_title}</span></a></figcaption>"
			elif link_caption == "none":
				caption_string = f"<figcaption><span class=\"figure-number\">{fig_num_format}</span><span class=\"figure-title\">{block_title}</span></figcaption>"
			
			if caption_before:
				processed_block = f"{caption_string}\n{processed_block}"
			else:
				processed_block = f"{processed_block}\n{caption_string}"
				
		processed_block = f"<figure{attrs}>{processed_block}</figure>"
		
		if retain_block == "comment":
			processed_block = f"<!--\n{block_match[0]}\n-->\n\n{processed_block}"
		elif retain_block == "indent":
			processed_block = f"{'\t'.join(('\t'+block_match[0].lstrip()).splitlines(True))}\n\n{processed_block}"
		
		last_fig_end = block_match.start() + len(processed_block) - pre_match_delta
		text = pre_match_text + processed_block + text[block_match.end():]
		figs_processed += 1
		block_match = block_pattern_obj.search(text, last_fig_end)
	
	if figs_processed > 0:
		print(f"Processed {figs_processed} FigureMark blocks.")
	else:
		print(f"No FigureMark blocks found.")
	
	return text
